feat(deadline): configure identity registration settings for deadline clients

jusiskin opened the pull request:
## Summary

This change makes RFDK automatically configure Deadline Secrets Management identity registration settings based on the RFDK constructs that are present in a CDK application.

The changes are broken out by construct below:

### `DeploymentInstance` (new)

This construct is an extraction/generalization of the CDK/CloudFormation pattern used by the `Repository` construct. It deploys an EC2 Auto-Scaling Group with a size of one. The Auto-Scaling Group is configured to require one CloudFormation success signal which is only sent after the user-data completes without error. The user data can optionally (by default enabled) self-terminate the instance by scaling down its own auto-scaling group to zero.

For now, this construct is being kept internal to RFDK and not exported. It was designed with the possibility of making it a public construct in the future.

### `SecretsManagementIdentityRegistration` (new)

The `SecretsManagementIdentityRegistration` construct is also introduced in this PR. It is responsible for configuring the user data (on a `DeploymentInstance`) to configure the identity registration settings. It adds IAM permissions and user-data to perform the following steps:

1.  download and install the Deadline Client
2.  configure Deadline to make a direct connection to the Deadline Repository
3.  run a python script that
    1.  fetches CIDR ranges for subnets involved
    2.  fetches existing identity registration settings
    3.  creates/updates/deletse identity registration settings so that they match what's desired

The main API of this construct is the `.addSubnetIdentityRegistrationSetting(...)` method which a caller can use to specify the subnets of the client, the desired secrets management [role](https://docs.thinkboxsoftware.com/products/deadline/10.1/1_User%20Manual/manual/secrets-management/deadline-secrets-management.html#assigned-roles) and [registration status](https://docs.thinkboxsoftware.com/products/deadline/10.1/1_User%20Manual/manual/secrets-management/deadline-secrets-management.html#registration-status).

### `RenderQueue` (modified)

This PR introduces a new `configureSecretsManagementAutoRegistration(...)` method to `IRenderQueue`. When called on the `RenderQueue`, it will lazily create `DeploymentInstance` and `SecretsManagementIdentityRegistration` constructs and delegate the work to the `SecretsManagementIdentityRegistration` construct. It also accepts a `dependent` prop which is used to enforce a CloudFormation dependency on a dependent resource of the caller's choosing.

The CloudWatch logs for the `DeploymentInstance` are sent to the `/renderfarm/ConfigureRepository` log group with the intention that the behavior of the deployment instance can be augmented in the future.

### `WorkerInstanceFleet` / `ConfigureSpotEventPlugin` / `UsageBasedLicensing` (modified)

The RFDK constructs that deploy Deadline Clients were modified to call the `configureSecretsManagementAutoRegistration(...)` method 

### `IVersion` / `VersionQuery` / `ThinkboxDockerRecipes.version` (modified)

The `IVersion` interface and classes that implement it were modified to return the path to the Deadline Client installer location. This was required in order for the `SecretsManagementIdentityRegistration` construct to be able to install the Deadline Client on instance deployed by the `DeploymentInstance`.

## Testing

Unit tests were updated to reflect all of the changes

Automatic identity registration of Deadline Clients was end-to-end tested by deploying the `All-In-AWS-Infrastructure-Basic` TypeScript example app with Thinkbox Usage-Based Licensing (UBL). Verified that:

*   Client identities were properly being assigned with the "Client" role and statuses by checking in a connected Deadline Monitor
*   Logs in the `renderfarm/ConfigureRepository` log group showed correct behavior
*   Clients were able to access Deadline-managed secrets. Thinkbox UBL requires the "Client" role to be assigned properly. Renders were successful.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

jusiskin committed:
feat(deadline): configure identity registration settings for Deadline client instances

jericht commented on packages/aws-rfdk/lib/deadline/scripts/python/configure_identity_registration_settings.py:
```diff
@@ -0,0 +1,634 @@
+        match = FILE_URI_RE.match( value )
+        if match:
+            named_groups = match.groupdict()
+            return FileSecret( arn=named_groups['FilePath'] )
```
```suggestion
            return FileSecret( file_path=named_groups['FilePath'] )
```

jusiskin replied:
The `FileSecret` was removed due to it being unused code.

horsmand commented on packages/aws-rfdk/lib/core/lib/deployment-instance.ts:
```diff
@@ -0,0 +1,287 @@
+ */
+export interface DeploymentInstanceProps {
+  /**
+   * The time CloudFormation should wait for the success signals before failing the create/update.
```
```suggestion
   * The amount of time that CloudFormation should wait for the success signals before failing the create/update.
```

horsmand commented on packages/aws-rfdk/lib/core/lib/deployment-instance.ts:
```diff
@@ -0,0 +1,287 @@
+
+  /**
+   * Properties for setting up the Deadline DeploymentInstance's LogGroup in CloudWatch
+   * @default - LogGroup will be created with all properties' default values to the LogGroup: /renderfarm/<construct id>
```
To keep a similar style to the rest of the props:

```suggestion
   * Properties for setting up the Deadline DeploymentInstance's LogGroup in CloudWatch
   *
   * @default the LogGroup will be created with all properties' default values to the LogGroup: /renderfarm/<construct id>
```

horsmand commented on packages/aws-rfdk/lib/core/lib/deployment-instance.ts:
```diff
@@ -0,0 +1,287 @@
+  /**
+   * The VPC that the instance should be launched in.
+   */
+  readonly vpc: IVpc;
```
I think for the most part we try to put the required fields at the top of the interface

jusiskin replied:
Done

horsmand commented on packages/aws-rfdk/lib/core/lib/deployment-instance.ts:
```diff
@@ -0,0 +1,287 @@
+
+    this.connections = this.asg.connections;
+
+    const logGroupName = props.logGroupName ?? id;
```
How obvious is it what the `id` belongs to if a user sees this log group in the CloudWatch console?

jusiskin replied:
I agree. I'm not a fan of how the log group name's suffix is derived from the construct ID. I'm following the pattern RFDK has followed so far. Here are a few example constructs that do the same:

*   `Repository`

    https://github.com/aws/aws-rfdk/blob/8845f4787cd7761f4b94213a7577ad6745d4246e/packages/aws-rfdk/lib/deadline/lib/repository.ts#L659
*   `RenderQueue`

    https://github.com/aws/aws-rfdk/blob/8845f4787cd7761f4b94213a7577ad6745d4246e/packages/aws-rfdk/lib/deadline/lib/render-queue.ts#L409-L412
*   `UsageBasedLicensing`

    https://github.com/aws/aws-rfdk/blob/8845f4787cd7761f4b94213a7577ad6745d4246e/packages/aws-rfdk/lib/deadline/lib/usage-based-licensing.ts#L553
*   `WorkerInstanceFleet`

    https://github.com/aws/aws-rfdk/blob/8845f4787cd7761f4b94213a7577ad6745d4246e/packages/aws-rfdk/lib/deadline/lib/worker-fleet.ts#L512-L521
*   `SpotEventPluginFleet`

    https://github.com/aws/aws-rfdk/blob/8845f4787cd7761f4b94213a7577ad6745d4246e/packages/aws-rfdk/lib/deadline/lib/spot-event-plugin-fleet.ts#L461-L474

horsmand commented on packages/aws-rfdk/lib/deadline/lib/render-queue.ts:
```diff
@@ -112,6 +123,19 @@ export interface IRenderQueue extends IConstruct, IConnectable {
+   *
+   * @param props Properties that specify thje configuration to be applied to the Deadline' Secrets Management identity
+   * registration settings. This specifies a VPC subnet and configures Deadline to automatically register identities of
+   * clients connecting from the subnet to a chosen Deadline Secrets Management role and status.
```
```suggestion
   * @param props Properties that specify the configuration to be applied to the Deadline' Secrets Management identity
   * registration settings. This specifies a VPC subnet and configures Deadline to automatically register identities of
   * clients connecting from the subnet to a chosen Deadline Secrets Management role and status.
```

horsmand commented on packages/aws-rfdk/lib/deadline/lib/render-queue.ts:
```diff
@@ -894,4 +948,47 @@ export class RenderQueue extends RenderQueueBase implements IGrantable {
+   * The construct that manages Deadline Secrets Management identity registration settings
+   */
+  private get identityRegistrationSettings(): SecretsManagementIdentityRegistration {
+    const secretsManagementIdentityRegistration = this.node.tryFindChild('SecretsManagementIdentityRegistration');
```
We could pull this string out like we did for the `deploymentInstance`.

jusiskin replied:
agreed. I've made the change

horsmand commented on packages/aws-rfdk/lib/deadline/lib/secrets-management-ref.ts:
```diff
@@ -0,0 +1,106 @@
+  readonly registrationStatus: SecretsManagementRegistrationStatus;
+
+  /**
+   * The role to assign to be assigned to the Deadline Client identities that connect from the specified VPC subnets.
```
```suggestion
   * The role to be assigned to the Deadline Client identities that connect from the specified VPC subnets.
```

horsmand commented on packages/aws-rfdk/lib/deadline/lib/secrets-management.ts:
```diff
@@ -0,0 +1,262 @@
+  readonly version: IVersion;
+}
+
+export class SecretsManagementIdentityRegistration extends Construct {
```
This class is missing a docstring. I believe we need resources deployed and security considerations sections? Not sure if it warrants a diagram, probably not.

jusiskin replied:
While this class is exported by the module, it is not exported from the package so it is internal. That being said, a docstring is still a good idea, so I'll add one.

jusiskin replied:
Added a TSDoc string. Let me know how it looks.

horsmand replied:
Looks good

horsmand commented on packages/aws-rfdk/lib/deadline/lib/render-queue.ts:
```diff
@@ -894,4 +948,47 @@ export class RenderQueue extends RenderQueueBase implements IGrantable {
+          deploymentInstance: this.deploymentInstance,
+          repository: this.repository,
+          renderQueueSubnets: this.props.vpc.selectSubnets(
+            this.props.vpcSubnetsAlb ?? RenderQueue.DEFAULT_VPC_SUBNETS_ALB,
```
Do you think using the default subnets justifies a warning message saying that their registration settings are likely too permissive?

jusiskin replied:
Good idea! I'll add this

jusiskin replied:
Applied this suggestion

horsmand commented on packages/aws-rfdk/lib/deadline/lib/secrets-management.ts:
```diff
@@ -0,0 +1,262 @@
+            `"${this.adminCredentials.secretArn}"`,
+          ],
+          // Subnets of the load balancer
+          (
+            props.renderQueueSubnets
+              .subnetIds
+              .map(subnetID => `--connection-subnet "${subnetID}"`)
+          ),
```
What are the extra parentheses for around this?

jusiskin replied:
This was done for horizontally aligning the commas in the outer list. What do you think? Does it help or hinder readability?

horsmand replied:
I'm not sure what you mean by aligning the commas in the outer list. It might align it with the array above it, but it doesn't look like it lines up with the `Array.from()` item below it. Also I'm not sure if there should be a dangling comma on the line with the `.map()`.

jusiskin replied:
The map doesn't require a comma dangle because it is the contents of bracket-ed expression.

I've made a pass at fixing the alignment. Let me know if it looks acceptable to you.

horsmand replied:
Yeah I like it better with the last item indented as well.

horsmand commented on packages/aws-rfdk/lib/deadline/lib/secrets-management.ts:
```diff
@@ -0,0 +1,262 @@
+          ),
+          Array.from(this.subnetRegistrations.entries())
+            // Each setting becomes a comma (,) separated string of fields
+            //   <SUBNET_ID>,<ROLE>,<REGISTRATIONSTATUS>
```
```suggestion
            //   <SUBNET_ID>,<ROLE>,<REGISTRATION_STATUS>
```

horsmand commented on packages/aws-rfdk/lib/deadline/lib/secrets-management.ts:
```diff
@@ -0,0 +1,262 @@
+              ].join(',');
+            })
+            // quote
+            .map(joinedSubnetArgValue => `--source-subnet "${joinedSubnetArgValue}"`),
```
Not sure what this comment is for.

jusiskin replied:
Oops, looks like was not updated after a change during implementation. I've revised the comment. Let me know what you think now.

horsmand replied:
I think it was straightforward enough not to need a comment, but it's fine as-is.

horsmand commented on packages/aws-rfdk/lib/deadline/lib/secrets-management.ts:
```diff
@@ -0,0 +1,262 @@
+  private runPythonScript(props: SecretsManagementIdentityRegistrationProps, localScriptFile: string) {
+    // The command-line arguments to be passed to the script that configures the Deadline identity registration
+    // settings
+    const scriptArgs = Lazy.list({
```
I had made a comment on the render queue about a warning if they're using the default subnets for the ALB, but maybe we could have a warning here instead if there is overlap between the subnets used for `--connection-subnet` and those in `--source-subnet`.

jusiskin replied:
Interesting decision. I'll think about this and try to arrive at the best approach. I agree that we can and should emit warnings in certain situations.

jusiskin replied:
I added this check among some others. What do you think?

horsmand replied:
Yep, you've covered everything that we recommend against. Looks good to me.

horsmand commented on packages/aws-rfdk/lib/deadline/scripts/python/configure_identity_registration_settings.py:
```diff
@@ -0,0 +1,607 @@
+        if subnet_id not in observed_subnets:
+            observed_subnets.add(subnet_id)
+        else:
+            raise ValueError(f"Subnet \"{subnet_id}\" is not unique")
```
This shouldn't be able to happen, but would it be better just to print a message and skip a duplicate?

jusiskin replied:
I'm not so sure... If two options for a subnet are passed with different effects, we aren't sure which one is actually desired/intended by the user. Since this is a user error and we can catch a mistake on their part that has impact on their security posture, I think it is safer to fail in such a scenario.

What do you think?

horsmand replied:
You have a good point that it's unclear which duplicate entry to keep. I was thinking that since we're only ever setting the registration settings to configure identities as `registered` with the `client` role, then the entries are identical, but it does have the capability to set entries to other values than `registered` and `client`. I think we leave it as-is.

horsmand commented on packages/aws-rfdk/lib/deadline/lib/secrets-management.ts:
```diff
@@ -0,0 +1,262 @@
+      bucket: props.version.linuxInstallers.client.s3Bucket,
+      bucketKey: props.version.linuxInstallers.client.objectKey,
+    });
+    props.deploymentInstance.userData.addCommands('set -x');
```
Will this cause it to print passwords anywhere? I'm guessing no since the Deadline Commands get run from in the python script, but just want to confirm.

jusiskin replied:
I've confirmed this does not by reviewing CloudWatch logs. Our repository direct connection script runs in a bash sub-process and that securely handles the password.

Also, the SM admin password used by `packages/aws-rfdk/lib/deadline/scripts/python/configure_identity_registration_settings.py` is put in the `DL_SM_PW` environment variable and we use the `--password env:DL_SM_PW` method to avoid passing the password with command-line arguments.

horsmand commented on packages/aws-rfdk/lib/deadline/scripts/python/configure_identity_registration_settings.py:
```diff
@@ -0,0 +1,607 @@
+        # On Linux, the Deadline Client installer creates a system-wide script to set the DEADLINE_PATH environment
+        # variable. Cloud-init does not load system environment variables. Cherry-pick the
+        # environment variable installed by the Deadline Client installer.
+        if not deadline_bin and os.path.exists(DL_ENV_SCRIPT_PATH):
```
Maybe we should rename this variable to have `LINUX` in the name and extract out the OSX path into a variable as well.

jusiskin replied:
done

horsmand commented on packages/aws-rfdk/lib/deadline/test/repository.test.ts:
```diff
@@ -949,6 +954,12 @@ test('repository configure client instance', () => {
   const regex = new RegExp(escapeTokenRegex('\'/tmp/${Token[TOKEN.\\d+]}${Token[TOKEN.\\d+]}\' \\"/mnt/repository/DeadlineRepository\\"'));
   expect(userData).toMatch(regex);
+
+  // TODO: Add test to assert the IAM instance profile is given read access to the database credentials secret
```
Is this something we need to do later?

jusiskin replied:
During implementation, I ran into an issue with being unable to connect to the database and thought that we'd missed this behavior. It actually turned out that the database credentials were stored only for the `ec2-user`. I've since added a code comment in the relevant spot, but since we were missing this test coverage, I've added it now.

horsmand commented on packages/aws-rfdk/lib/deadline/test/secrets-management.test.ts:
```diff
@@ -0,0 +1,747 @@
+      });
+
+      // eslint-disable-next-line jest/no-commented-out-tests
+      // test('normalizes subnet ID order', () => {
```
How come this test is commented out? I can't understand why subnet ID order matters.

jusiskin replied:
Initially I'd thought the order needed to be normalized because I'd thought that changes to the CDK code structure that change the call order to `RenderQueue.configureSecretsManagementAutoRegistration(...)` would cause the user data to change and trigger the configuration to happen again unnecessarily.

This is an optimization and not strictly required for the feature to work correctly. It also turned out that ordering subnet IDs is not always possible when they are unresolved CDK tokens. For this reason, I removed that from the implementation.

I commented out the test and forgot to remove it completely before pushing it up. The next revision will have this test removed.

horsmand commented on packages/aws-rfdk/lib/core/test/deployment-instance.test.ts:
```diff
@@ -0,0 +1,763 @@
+  let target: DeploymentInstance;
+  let deploymentInstanceProps: DeploymentInstanceProps;
+
+  beforeEach(() => {
```
Running this setup method before each test and having tests only assert on one field seems like a heavyweight way to do testing. Is there any opportunity to switch some of these `beforeEach()` setup methods to `beforeAll()` methods?

jusiskin replied:
Keep in mind that using `beforeAll()` to create partial CDK construct trees means that tests would share the construct tree. A few reasons below why this is problematic/challenging:

1.  some tests/fixtures mutate the construct tree, so when a test calls `expectCDK(...).to(...)`, then we might get results that are confounded by inter-test pollution
2.  we'll run into duplicate construct ID errors unless care is taken to ensure unique IDs between each test code path.
3.  anyone working on these tests in the future will need to be aware that the tests use a shared construct tree. The difference between `beforeEach` and `beforeAll` is subtle and someone adding a test might not realize this and waste time debugging their tests

I refactored the tests to use `beforeAll(...)` everywhere and then timed 20 exeuctions of each implementation with:

```sh
for x in $(seq 1 20); do
    yarn jest lib/core/test/deployment-instance.test.js  |& egrep "^Time:" | awk '{ print $2 }'
done
```

Here are the results on my dev machine:

| Iteration | `beforeAll(...)` (seconds) | `beforeEach(...)` (seconds) |
| --------- | ---------------- | ----------------- |
| 1 | 4.946 | 4.805 |
| 2 | 4.568 | 4.779 |
| 3 | 4.538 | 4.854 |
| 4 | 4.506 | 4.869 |
| 5 | 4.525 | 4.83 |
| 6 | 4.668 | 4.85 |
| 7 | 4.556 | 5.037 |
| 8 | 4.542 | 4.761 |
| 9 | 4.764 | 4.961 |
| 10 | 4.582 | 4.959 |
| 11 | 4.94 | 4.81 |
| 12 | 4.666 | 4.847 |
| 13 | 4.917 | 4.803 |
| 14 | 4.494 | 4.894 |
| 15 | 4.559 | 4.851 |
| 16 | 4.599 | 5.214 |
| 17 | 4.639 | 4.816 |
| 18 | 4.733 | 4.834 |
| 19 | 4.987 | 5.121 |
| 20 | 4.533 | 4.88 |

Granted, these times are likely governed by loading up the node, jest, and the dependent code, JIT compiling it all, etc.. 

The statistical distributions for each are:

| Implementation | mean | stddev |
| --- | --- | --- |
| `beforeAll` | 4.6631 | 0.1627701 |
| `beforeEach` | 4.88875 | 0.1167336 |

The difference between means is ~220ms. Compounded that can add up. 100 runs of that test would mean 22seconds of saved compute in the `beforeAll` case.

Do you think that this is worth the trade-offs mentioned above?

jusiskin replied:
I've refactored to use `beforeAll()` where possible

horsmand reviewed (changes requested):
Overall this looks really good, none of my comments should require changing too much in the overall architecture of this. I think the deployment instance could be a really powerful tool.

jusiskin committed:
Apply suggestions from code review

Co-authored-by: David Horsman <56004724+horsmand@users.noreply.github.com>

jericht commented on packages/aws-rfdk/lib/core/lib/deployment-instance.ts:
```diff
@@ -0,0 +1,288 @@
+  readonly logGroupName?: string;
+
+  /**
+   * Properties for setting up the Deadline DeploymentInstance's LogGroup in CloudWatch
```
```suggestion
   * Properties for setting up the DeploymentInstance's LogGroup in CloudWatch
```

nit: looks like a leftover `Deadline` we can remove since this class is generalized

jusiskin replied:
good catch - fixed

jericht commented on packages/aws-rfdk/lib/core/lib/deployment-instance.ts:
```diff
@@ -0,0 +1,288 @@
+    this.asg = new AutoScalingGroup(this, 'ASG', {
+      instanceType: props.instanceType ?? InstanceType.of(InstanceClass.T3, InstanceSize.SMALL),
+      keyName: props.keyName,
+      machineImage: props.machineImage ?? new AmazonLinuxImage(),
```
```suggestion
      machineImage: props.machineImage ?? MachineImage.latestAmazonLinux(),
```

nit: I think calling `latestAmazonLinux()` here makes this is a bit easier to read since the default constructor for `AmazonLinuxImage`  [uses the latest one anyway](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-ec2.AmazonLinuxImage.html), which wasn't obvious to me at first glance. What do you think?

jusiskin replied:
sure! that reads better - applied now.

jericht replied:
thanks!

jericht commented on packages/aws-rfdk/lib/core/lib/deployment-instance.ts:
```diff
@@ -0,0 +1,288 @@
+    }
+  }
+
+  public get osType() {
```
```suggestion
  /** 
   * @inheritdoc
   */
  public get osType() {
```

nit: This getter and the one below are inheriting their docstring right?

jusiskin replied:
VSCode seems to think so when I hover over them, but I'm not sure whether our auto-generated docs will behave the same. I'll add it to be explicit.

jericht commented on packages/aws-rfdk/lib/core/lib/deployment-instance.ts:
```diff
@@ -0,0 +1,288 @@
+   *   - The cloud-init log
+   *
+   * @param asg The auto-scaling group
+   * @param logGroupProps The properties for LogGroupFactory to create or fetch the log group
```
nit: missing `@param` tag for `groupName`. This is a private method though, so not necessary. We can even leave out this entire comment if we want.

jusiskin replied:
I opted to add in the `@param`. Let me know what you think

jericht replied:
Looks good, thanks

kozlove-aws commented on packages/aws-rfdk/lib/deadline/lib/secrets-management-ref.ts:
```diff
@@ -0,0 +1,106 @@
+ *
+ * See https://docs.thinkboxsoftware.com/products/deadline/10.1/1_User%20Manual/manual/secrets-management/deadline-secrets-management.html#registration-status
+ */
+export enum SecretsManagementRegistrationStatus {
```
What the point to have these statuses? Do we expect users will use something except `REGISTERED`?

jusiskin replied:
I made all possible identity registration statuses available that Deadline supports. RFDK users should not need to interact with this part of the API, since the constructs will automatically work together to configure the identity registration settings.

The only way a user could use this API is if they called `RenderQueue.configureSecretsManagementAutoRegistration(...)` directly. On that note, I'll document that fact on the method's TSDoc header.

kozlove-aws replied:
I mean, since we do not need others statuses we can just use constant `REGISTERED` in python script and do not ask for status in `SubnetIdentityRegistrationSettingsProps` unless we want to register settings with different statuses.

jusiskin replied:
Added a TSDoc string explaining that this API should not used by RFDK end-users unless you have a specific need. Let me know what you think.

kozlove-aws replied:
Hm. Where is this new TSDoc string?

jusiskin replied:
https://github.com/aws/aws-rfdk/pull/576/files#diff-7087a85eeaddfc36b08d14a6bb1536c2c5a6291bf28d1c9ec2a65e4597c6ec09R134-R136

kozlove-aws commented on packages/aws-rfdk/lib/deadline/scripts/python/configure_identity_registration_settings.py:
```diff
@@ -0,0 +1,607 @@
+
+    prior_rfdk_settings = get_rfdk_registration_settings(dl_secrets)
+
+    delete_removed_settings(dl_secrets, prior_rfdk_settings, connection_subnets, source_subnets)
```
Am I understand right that settings that created manually will be removed?

jusiskin replied:
The code uses a RFDK-specific convention for naming the identity registration settings. The settings have a name convention of:

```
RfdkSubnet|<CONNECTING_SUBNET_ID>|<SOURCE_SUBNET_ID>
```

This code will only remove settings that:

1.  match the above naming convention
2.  do not match the connecting/source subnets supplied in the command-line arguments 

So the only scenario where it would delete a manually created identity registration setting is if a user created one that matches the RFDK naming convention.

kozlove-aws replied:
Right, I missed that we have strong name filter for these settings.

kozlove-aws reviewed (commented):
Great job. Just a couple moments that I want clarify.

jericht commented on packages/aws-rfdk/lib/core/test/deployment-instance.test.ts:
```diff
@@ -0,0 +1,763 @@
+
+      });
+
+      test('Uses created Security Group', () => {
```
Is this test necessary? Based on our code in `DeploymentInstance`, it doesn't look like we're doing anything fancy with the security group, so it seems like this test is actually testing the CDK construct, not ours.

Same question for the `depends on policy` test below

jusiskin replied:
If we remove the test, then we don't have anything asserting what security group the synthesized ASG uses.

I understand that we defer that logic to CDK, but shouldn't we still assert that we are using CDK's security group code? The other option for that would be mocking, and I opted for `cdk-assert`.

What do you think?

jericht replied:
That's a good point. I wanted to make sure we weren't unnecessarily testing CDK code, but I guess this is just one of the nuances when working with CDK since we're asserting the generated CFN template and not directly asserting the logic in our code.

It doesn't hurt to have these extra tests in, so let's keep them. Resolving these comments

jericht commented on packages/aws-rfdk/lib/core/test/deployment-instance.test.ts:
```diff
@@ -0,0 +1,763 @@
+    });
+
+    describe('Security Group', () => {
+      test('creates Security Group in the desired VPC', () => {
```
Same question here, is this testing CDK logic? Maybe we can omit this test

jericht commented on packages/aws-rfdk/lib/core/test/deployment-instance.test.ts:
```diff
@@ -0,0 +1,763 @@
+        ) as iam.CfnRole;
+      });
+
+      test('creates an instance profile', () => {
```
Same thing here, do we need this test and the `creates a role that can be assumed by EC2` test below?

jericht commented on packages/aws-rfdk/lib/core/test/deployment-instance.test.ts:
```diff
@@ -0,0 +1,763 @@
+              {
+                Action: 'autoscaling:UpdateAutoScalingGroup',
+                Condition: {
+                  StringEquals: { 'autoscaling:ResourceTag/resourceLogicalId': cdk.Names.uniqueId(target) },
```
Do you think it would be helpful to leave a comment here saying that the key `autoscaling:ResourceTag/resourceLogicalId` is set by the `DeploymentInstance` class? Specifically the `resourceLogicalId` part. At first glance, it kind of looks like this could be a tag created by AWS or CDK and not our code.

jusiskin replied:
I added this now. Let me know what you think

jericht replied:
I don't see any changes around here, did you forget to commit this change?

jericht replied:
Nevermind, I missed it elsewhere in the diff, resolving :)

jusiskin replied:
Oops. I updated here:

https://github.com/aws/aws-rfdk/pull/576/files#diff-f29bdea96151f22e55d5bf9b6dacb16ac189706acb83b3607d8c4dbaea2c0024R701

but not at this line as well. I'll add that....

jericht commented on packages/aws-rfdk/lib/core/test/deployment-instance.test.ts:
```diff
@@ -0,0 +1,763 @@
+      });
+    });
+
+    describe('CloudWatch Agent config SSM paramter', () => {
```
```suggestion
    describe('CloudWatch Agent config SSM parameter', () => {
```

jusiskin replied:
fixed

jusiskin committed:
chore(deadline): apply CR feedback

- reorder DeploymentInstance props
- document internal use of RenderQueue.configureSecretsManagementAutoRegistration
- move SecretsManagementIdentityRegistration construct ID to constant
- Add TSDoc header for SecretsManagementIdentityRegistration construct
- Factor out constant for MacOS Deadline path file
- Remove artifacts from FileSecret in configure_identity_registration_settings.py
- fixes to encoding, logging and error-handling of deadline path on MacOS

jericht commented on packages/aws-rfdk/lib/core/test/deployment-instance.test.ts:
```diff
@@ -0,0 +1,763 @@
+        '/renderfarm/foo',
+      ],
+      [
+        { logGroupName: 'bar' },
```
Can we remove this test case? Looks like this has the same coverage as the previous `foo` case.

jusiskin replied:
Removed

jericht commented on packages/aws-rfdk/lib/core/test/deployment-instance.test.ts:
```diff
@@ -0,0 +1,763 @@
+  });
+
+  // GIVEN
+  describe.each<[InstanceType | undefined]>([
```
Does this need to be its own `describe`  block with multiple test cases? We have the default test `uses t3.small` which asserts the default instance type, and the last two test cases cover the same code paths. Can we simplify this into a single test that asserts that the ASG uses the provided instance type?

jusiskin replied:
yep good call. less code is better! done

jericht replied:
awesome, thanks!

jericht commented on packages/aws-rfdk/lib/core/test/deployment-instance.test.ts:
```diff
@@ -0,0 +1,763 @@
+      ]));
+    });
+
+    test('is not grante IAM permissions to scale the Auto-Scaling Group', () => {
```
```suggestion
    test('is not granted IAM permissions to scale the Auto-Scaling Group', () => {
```

jusiskin replied:
Fixed - thanks for catching

jericht commented on packages/aws-rfdk/lib/core/test/deployment-instance.test.ts:
```diff
@@ -0,0 +1,763 @@
+  });
+
+  // GIVEN
+  describe.each<[string]>([
```
Since we only have a single case here, do we need the `.each` from the `describe`? I don't think we will need to add more cases here in the future. Maybe we can use the value in the description text or rename it to `.executionTimeout is specified` or something

jusiskin replied:
Good feedback. I've taken a pass at it. Let me know if it looks okay.

jericht replied:
looks great, thanks!

jericht commented on packages/aws-rfdk/lib/deadline/lib/render-queue.ts:
```diff
@@ -112,6 +123,19 @@ export interface IRenderQueue extends IConstruct, IConnectable {
+   * See https://docs.thinkboxsoftware.com/products/deadline/10.1/1_User%20Manual/manual/secrets-management/deadline-secrets-management.html#identity-management-registration-settings-ref-label
+   * for details
+   *
+   * @param props Properties that specify the configuration to be applied to the Deadline' Secrets Management identity
```
```suggestion
   * @param props Properties that specify the configuration to be applied to the Deadline Secrets Management identity
```

jusiskin replied:
Good catch - applied

jericht commented on packages/aws-rfdk/lib/deadline/lib/render-queue.ts:
```diff
@@ -894,4 +948,47 @@ export class RenderQueue extends RenderQueueBase implements IGrantable {
+      return new DeploymentInstance(this, CONFIGURE_REPOSITORY_CONSTRUCT_ID, {
+        vpc: this.props.vpc,
+        vpcSubnets: {
+          subnetType: SubnetType.PUBLIC,
```
Maybe a naive question, but why do we need to put this in a public subnet instead of a private one?

jusiskin replied:
Not naive at all - in fact this is a great catch! We need to lock down this instance since it has full access to:

*   Repository file-system
*   Repository database
*   Database credentials
*   Secrets management admin credentials secret in AWS Secrets Manager

This was changed to public for speeding up my dev/test cycle iterations and it looks like I forgot to set it back. I've changed it to use the same subnets passed into the render queue's props. I've added a test to make sure this doesn't regress.

jericht replied:
Awesome, thanks for clearing that up for me and adding the additional testing

jericht commented on packages/aws-rfdk/lib/deadline/lib/render-queue.ts:
```diff
@@ -894,4 +948,47 @@ export class RenderQueue extends RenderQueueBase implements IGrantable {
+    } else if (deploymentInstanceNode instanceof DeploymentInstance) {
+      return deploymentInstanceNode;
+    } else {
+      throw new Error(`Unexpected type for ${deploymentInstanceNode.node.path} (${typeof(deploymentInstanceNode)})`);
```
This error will most likely never happen, but would it help to put the expected type here?

```suggestion
      throw new Error(`Unexpected type for ${deploymentInstanceNode.node.path}. Expected ${DeploymentInstance.name}, but found ${typeof(deploymentInstanceNode)}.`);
```

jusiskin replied:
Applied

jericht commented on packages/aws-rfdk/lib/deadline/lib/render-queue.ts:
```diff
@@ -894,4 +948,47 @@ export class RenderQueue extends RenderQueueBase implements IGrantable {
+    } else if (secretsManagementIdentityRegistration instanceof SecretsManagementIdentityRegistration) {
+      return secretsManagementIdentityRegistration;
+    } else {
+      throw new Error(`Unexpected type for ${secretsManagementIdentityRegistration.node.path} (${typeof(secretsManagementIdentityRegistration)})`);
```
Same thing here, should we add the expected type?
```suggestion
      throw new Error(`Unexpected type for ${secretsManagementIdentityRegistration.node.path}. Expected ${SecretsManagementIdentityRegistration.name}, but found ${typeof(secretsManagementIdentityRegistration)}.`);
```

jusiskin replied:
Applied

jericht commented on packages/aws-rfdk/lib/deadline/lib/secrets-management-ref.ts:
```diff
@@ -0,0 +1,106 @@
+   * [Remote Connection Server](https://docs.thinkboxsoftware.com/products/deadline/10.1/1_User%20Manual/manual/remote-connection-server.html#remote-connection-server-ref-label)
+   * application. The Server role is granted to a registered machine by an administrator in the
+   * [Monitor UI](https://docs.thinkboxsoftware.com/products/deadline/10.1/1_User%20Manual/manual/secrets-management/deadline-secrets-management.html#identity-management-assigning-ref-label).
+   * In order to encrypt and decrypt secrets, the master key must be assigned to the Server by and Administrator user
```
```suggestion
   * In order to encrypt and decrypt secrets, the master key must be assigned to the Server by an Administrator user
```

jusiskin replied:
thanks - fixed

jericht commented on packages/aws-rfdk/lib/deadline/lib/secrets-management.ts:
```diff
@@ -0,0 +1,262 @@
+  readonly registrationStatus: SecretsManagementRegistrationStatus;
+
+  /**
+   * The role to assign to be assigned to the Deadline Client identities that connect from the specified VPC subnets.
```
```suggestion
   * The role to be assigned to the Deadline Client identities that connect from the specified VPC subnets.
```

jusiskin replied:
Fixed

jericht commented on packages/aws-rfdk/lib/deadline/lib/secrets-management.ts:
```diff
@@ -0,0 +1,262 @@
+} from '../../core';
+import { DeploymentInstance } from '../../core/lib/deployment-instance';
+
+interface RegistrationSettingStack {
```
nit: missing docstring. This is not a necessary change since this interface isn't exported, but might be helpful to people that read this, especially since all the fields below mention subnets, so it'd be nice to have something here to glue them all together

jusiskin replied:
This led me to realize that the `stack` field is unused. I renamed the struct `RegistrationSettingEffect` such it only contains the output fields of a Deadline Secrets Management identity registration setting (role and status).

Since the structure is independent of subnets, I'll keep the TSDoc string general. Added now so let me know what you think

jericht commented on packages/aws-rfdk/lib/deadline/lib/stage.ts:
```diff
@@ -228,4 +229,39 @@ export class Stage {
+        this.dirPath,
+        'bin',
+        listing[0],
+      );;
```
```suggestion
      );
```

jusiskin replied:
done

jericht commented on packages/aws-rfdk/lib/deadline/test/usage-based-licensing.test.ts:
```diff
@@ -103,29 +110,71 @@ describe('UsageBasedLicensing', () => {
+
+    test.each<[SubnetSelection]>([
+      [{
+        subnetType: SubnetType.PUBLIC,
```
Do we need to use `test.each` here? Can we simplify this to just be a single test?

jusiskin replied:
I've moved away from `.each` for this single test case.

jericht replied:
The diff is still showing this as using `test.each`, but this is a super minor comment, so resolving

jusiskin committed:
feat(deadline): warn about dedicated subnets during synthesis

jericht commented on packages/aws-rfdk/lib/deadline/scripts/python/configure_identity_registration_settings.py:
```diff
@@ -0,0 +1,607 @@
+        for registration_setting in dl_secrets.run_json('GetLoadBalancerIdentityRegistrationSettings')
+    ]
+
+    print('all registration settings = ', end='')
```
Are these debug statements that we should remove? or do we want to keep this?

jusiskin replied:
These are pretty sparse, and I think they might be helpful for troubleshooting if any issues arise. What do you think though?

horsmand replied:
I don't see any harm in them, this file is only ran once per deployment so the amount of logs is pretty small and they are helpful.

jericht replied:
Yeah, totally fine to leave them in if they are intended logging messages, just wanted to make sure they weren't left in from development. Resolving

jericht reviewed (changes requested):
Looks great! Just have a few minor comments

jusiskin committed:
chore(deadline): optimize DeploymentInstance tests

jusiskin committed:
test(deadline): ensure Repository.configureClientInstance grants db secret read access

jusiskin committed:
chore(deadline): clean up code that generates parameters for identity registration configuration

jusiskin committed:
fix(deadline): deploy RenderQueue DeploymentInstance to specified subnets or private

jusiskin committed:
chore(deadline): apply CR feedback

- TSDoc improvements
- Use MachineImage.latestAmazonLinux
- Remove unnecessary test parameterization
- Add code comments
- Fix typos/punctuation/redundant syntax

jusiskin committed:
fix(deadline): typo in method name

jusiskin committed:
chore(deadline): add test code comment

jusiskin committed:
fix(deadline): fixed method call signature mismatch

jericht commented on packages/aws-rfdk/lib/deadline/test/secrets-management.test.ts:
```diff
@@ -586,111 +587,30 @@ describe('SecretsManagementIdentityRegistration', () => {
-      //   }];
-      //   expect(resolvedCalls).toContainEqual(expectedCall);
-      // });
+      test('warns about dedicated subnets when ALB and source subnets match', () => {
```
```suggestion
      test('warns about dedicated subnets when render queue ALB and source subnets match', () => {
```

this is a really minor nit though, so I won't hold the PR back for it

jusiskin replied:
Fixed

jericht commented on packages/aws-rfdk/lib/deadline/test/usage-based-licensing.test.ts:
```diff
@@ -121,6 +121,45 @@ describe('UsageBasedLicensing', () => {
   }
 
+  describe('vpcSubnets', () => {
+    describe('specified', () => {
```
Do we need `describe` blocks for the `specified` and the other `not specified` blocks? Since we only have one test case in each block, maybe we can just have to tests in this `vpcSubnets` block. What do you think?

Another really minor comment though, not holding back the PR for it.

jusiskin replied:
I've flattened the test hierarchy. Let me know what you think

jericht replied:
Looks good!

jericht commented on packages/aws-rfdk/lib/deadline/test/worker-fleet.test.ts:
```diff
@@ -1943,3 +1944,71 @@ test('worker fleet does not signal when zero minCapacity', () => {
+    props = {
+      renderQueue,
+      vpc,
+      workerMachineImage: new GenericWindowsImage({
```
```suggestion
      workerMachineImage: new GenericWindowsImage(),
```

We can remove the empty props `{}` here

jusiskin replied:
This is actually the AMI region map and its required. I've collapsed it onto a single line though which makes it looks a bit nicer

jericht replied:
Ah, I should've double checked the API. Thanks

jericht commented on packages/aws-rfdk/lib/deadline/test/worker-fleet.test.ts:
```diff
@@ -1943,3 +1944,71 @@ test('worker fleet does not signal when zero minCapacity', () => {
+    };
+  });
+
+  describe('vpc subnets not specified', () => {
```
Same comment here, do we need these to be `describe` blocks or can they just be `test`?

jusiskin replied:
I've flattened this now

jericht reviewed (dismissed):
LGTM!

I left a few minor comments, but it's mostly tiny formatting/organization in the unit tests, so not holding back the PR for it.

horsmand commented on packages/aws-rfdk/lib/deadline/test/configure-spot-event-plugin.test.ts:
```diff
@@ -968,4 +969,100 @@ describe('ConfigureSpotEventPlugin', () => {
+      groupName = 'group_name1';
+    });
+
+    describe('a fleet without vpcSunbets', () => {
```
```suggestion
    describe('a fleet without vpcSubnets', () => {
```

horsmand reviewed (dismissed):
I'm in the same boat as Jericho. This looks good as-is. I'm going to give my approval but leave it for Josh to look over the comments and decide whether to fix them or merge.

jusiskin committed:
chore(deadline): address PR feedback

- flatten jest test heirarchies
- fix typos
- clarify test name
